man/man3/table.3
.TH TABLE 3NEMO "5 May 2022"

.SH "NAME"
table_open, table_row, table_md2, table_close - table manipulation routines
.PP
get_atable, get_ftable, get_line, parse, strinsert \- legacy table routines

.SH "SYNOPSIS"
Note the API below has not been implemented for functions with a preceding dash!

.nf
.B #include <table.h>
.B #include <mdarray.h>
.B #include <extstring.h>
.PP
.B table  *table_open(stream instr, int mode);
.B table  *table_open1(stream instr, int mode, int nlines);
.B void    table_close(tableptr tptr);
.B size_t  table_nrows(tableptr tprt);
.B size_t  table_ncols(tableptr tprt);
.B string  table_row(tableptr tptr, int row);
.PP

.B string table_line(table *t);
.B ssize_t table_line1(table *t, char **line, size_t *linelen, int newline);
.PP
.B mdarray2 table_md2rc(table *t);
.B mdarray2 table_md2cr(table *t);
.B - string *table_comments(table *t);
.B void table_reset(table *t);
.B void table_close(table *t);
.PP
.B - void table_set_valid_rows(int nrows, int *rows)
.B - void table_set_valid_cols(int ncols, int *cols)
.PP
.B - int table_next_row(table *t)   
.B - int table_next_rows(table *t)  
.B - int table_next_rowi(table *t)
.B - int table_next_rowr(table *t)
.PP
.B - string table_cols(table *t, int col)
.B - int  table_coli(table *t, int col)
.B - real table_colr(table *t, int col)
.PP
.B - string *table_colsp(table *t, int col) 
.B - int *table_colip(table *t, int col)
.B - real *table_colrp(table *t, int col)
.PP
.B string table_row(table *t, int row)
.B string *table_rowsp(table *t, int row)
.B - int *table_rowip(table *t, int row)
.B - real *table_rowrp(table *t, int row)
.PP
.B - void table_set_ncols(int ncols)
.PP
.B table *table_cat(int ntable, table *tptr, int mode)
.PP
.I Legacy: (some of these might be deprecated in future)
.PP
.B int get_atable(strean instr,int ncol,int *colnr,real *coldat,int ndat)
.PP
.B int get_ftable(stream instr,int ncol,int *colpos,string *colfmt,real *coldat,int ndat) 
.PP
.B int get_line (string instr, char *line) // deprecated now
.PP
.B int parse(int linenr, char *line, double *dat, int ndat)
.PP
.B int strinsert(char *a, char *b, int n)
.PP
.B int iscomment(char *line)
.PP
.B void sanitize(char *line)
.fi

.SH "USING TABLES"

Tables can be arbitrarely large, but like top level of NEMO structured files,
do need to fit in memory. There can be some exceptions to this,
and certain applications will stream data that do not fit in
memory.

.PP
The particular table dialect should be detected automatically in
files, and number of columns and rows can be arbitrarely high.
\fItabgen(1NEMO)\fP
has methods to make really large tables in both "column" and "row" space
for testing purposes.

.PP
When a table comes from a pipe, obviously it cannot seek, however
using the new table I/O the full table is read into memory, initially
using a linked list, and can then be addressed line by line in random
row fashion.

.PP
An exception could be made where the table can be read in blocks of say 100
lines or so?

.PP
Comment lines: there are too many conventions for this, but we need to cover
them all.   A line starting with
.nf

.B #
.B !
.B ;
.B /
.B \e
.B |
.B C

.fi
have all been used for comments.  In addition comments have been used
at the end of a line, e.g.
.nf

.B     1 2 3   # this is a comment

.fi
but see \fItable(5NEMO)\fP for the official rules.
     

.SH "DESCRIPTION"
This new \fItable\fP interface was introduced in Summer 2020 to create a more uniform and scaleable
table ASCII system. This will have no limit on number of columns, line length or number or rows,
perhaps at a slight cost of performance. It should also detect automatically if tables have
Space Separated Values (SSV), Comma Separated Values (CSV) or Tab Separated Values (TSV).
.PP
.B table_open
opens a file for reading. The returned \fItable*\fP pointer is used in all subsequent table_ routines.
The
.B mode
controls how many lines from the table are allowed in internal buffers. A value of \fB0\fP
means the whole table will be read in memory, a value of \fB1\fP will read the table line
by line, controlled by the user (see \fBtable_line\fP below). Performance will be better (?)
if tables are read line by line (mode=1), or at least not occupy memory for the whole table.
Values larger than 1 are planned to hold small buffers of rows. Normally a table will
be split in a \fI"header"\fP (comment lines) and \fI"data"\fP (rows of data), but
with special \fImode=-1\fP all lines are treated equal and can be obtained via
\fBtable_row()\fP.
.PP
.B table_open1
is kept for compatibility with older softwhere where the maximum number of lines
.B nlines
is given. 
It normally is only
needed when the input file is a pipe and the whole file needs to be read, which is now supported.
.PP
.B table_line
will read the next line from the table stream.  If the file had been opened in \fBmode=0\fP all
lines have been read,  and \fBtable_line\fP would return NULL. Note that the returned string
is 0-terminated, not newline terminated as \fIgetline(3)\fP would do.

.B table_line1
will read the next line from the table stream using \fIgetline(3)\fP, but depending on the setting of
\fInewline\fP, the string can still contain a newline character. To best mimic the behavior
of \fIgetline(3)\fP, \fInewline=1\fP needs to be set.

.PP
.B table_md2cr, table_md2rc
are shortcut functions to convert an ascii table immediately into a two dimensional \fImdarray(3NEMO)\fP
data, for the [col][row] or [row][col] notation resp.
With
.PP
Any comment lines at the start of the file will saved in a special
\fIcomment\fP set of lines, which can be extracted with
.B table_comments.
Finally
.B table_close
access to the table can be closed and any associated memory will be freed. In addition
.B table_reset
can be used to reset array access (more on that later), in the case it needs to be re-read.
For arrays that are processed in streaming mode (e.g. \fIfilename="-"\fP) this will result in an error.
.PP
Once a table has been fully read into memory,
.B table_nrows
returns the number of (data, i.e. non-comment) rows (assuming non-streaming), and 
.B table_ncols
the number of columns. By using
.B table_set_valid_rows
and/or
.B table_set_valid_cols
rows and/or columns can be selected for conversion, and this will also define the new value for
.I nrows
and
.I ncols.
When
.B table_reset
is called, these values are reset to their original value.
.PP
If the table is parsed line by line, some routines will not be accessible, since the table is not
in memory. 
.PP
Using
.B table_next_row
a new line can be read. This will return -1 upon end of file, 0 when the line is blank or contains
no data, though could contain comments (e.g. lines with # ! or ;), and 1 when a line
was read. No parsing will be done. If parsing is done, the line will be tokenized
in identical types (\fBs\fPtring, \fBi\fPnt or \fBr\fPeal), with resp.
.B table_next_rows
,
.B table_next_rowi,
or
.B table_next_rowr.
The last line is always stored internally, and a pointer to the string can be retrieved with
.B table_line
for more refined user parsing.
.PP
Depending on with which of the three types the line was parsed, column elements can be retrieved with
.B table_cols,
.B table_coli,
or
.B table_colr.
and if the whole table was available in memory, columns can also be retrieved in full via
.B table_colsp,
.B table_colip,
or
.B table_colrp
.PP
The currently parsed row can in full be retrieved with (again, depending on type)
.B table_rowsp,
.B table_rowip,
or
.B table_rowrp
where the row number is ignored if the table is parsed row by row.
.PP
Possible future routines are
.B table_set_ncols
to cover the case where a row can span multiple lines. By default each line is a row in the table.
.PP
Given a number of tables, the
.B table_cat
function will catenate them. mode=0 will catenate them \fIvertically\fP, i.e. append the rows, keeping the
same number of columns, whereas mode=1 will catenate them \fIhorizontally\fP, keeping the number of
rows, but increasing the number of columns. These are similar to the unix programs
\fIcat(1)\fP and \fIpaste(1)\fP resp. It is currently considered an error if the tables are
not conformant in size.

.PP
The original \fIlegacy table routines\fP remain available, though they should implement the new API,
as it better deals with tables of unknown size in a pipe.
.PP
Both \fIget_atable\fP
and \fIget_ftable\fP parse an ascii table, pointed by the \fIinstr\fP stream,
into \fIncol\fP columns and up to \fIndat\fP rows of \fBreal\fP numbers
into memory. The input table may contain comment lines, as well as columns
which are not numbers. Badly parsed lines are simply skipped.
Other common parameters to both routines 
are \fIcoldat, ncol\fP and \fIndat\fP:
\fIcoldat\fP is an array of \fIncol\fP pointers to 
previously allocated data, each of them \fIndat\fP \fBreal\fP elements.
The number of valid rows read is then returned. If this number is negative,
it means more data is available, but could not be read because 
\fIndat\fP was exhausted. Upon the next call \fIndat\fP must be set to
negative, to recover the last line read on the previous call, and continue
reading the table without missing a line. CAVEAT: this only works if
\fIinstr\fP has not changed. 
.PP
\fIget_atable\fP parses the table in free format.
\fIcolnr\fP an array of length \fIncol\fP
of the column numbers to read (1 being the first column),
If any of the \fIcolnr\fP is 0, it is 
interpreted as referring to the line number in the
original input file (including/excluding comment and empty lines), 
1 being the first line, and the
corresponding entry in \fIcoldat\fP is set as such.
Columns are separated by whitespace or commas.
.PP
\fIget_ftable\fP parses the table in fixed format.
\fIcolpos\fP is an array with 
positions in the rows to start reading (1 being the first position),
\fIcolfmt\fP an array of pointers to the format string
used to parse a \fBreal\fP number
(note \fBreal\fP normally requires \fI%lf\fP).
If any of the \fIcolpos\fP is 0, it is interpreted as referring to the line 
number in the
original input file (including comment lines), 1 being the first line, and the
corresponding entry in \fIcoldat\fP is set as such.
.PP
The \fIget_line(3)\fP gets the next line from a stream \fIinstr\fP, stored
into \fIline\fP. It returns the length of the string read, 0 if end of file.
This routine is deprecated, the standard \fIgetline(3)\fP should be used.
.PP
\fIparse\fP parses the character string in \fIline\fP into the double array
\fPdat\fP, which has at most \fIndat\fP entries. Parsing means that 
\fB%n\fP refers to column \fBn\fP in the character string (\fBn\fP must
be larger than 0. Also \fB%0\fP may be referenced, meaning the current
line number, to be entered in the argument \fIlinenr\fP.
.PP
\fIstrinsert\fP inserts the string \fIb\fP into \fIa\fP, replacing \fIn\fP
characters of \fIa\fP.
.PP
\fIiscomment\fP returns 1 if the line appears to be a comment
(starts with ';', '#', '!' or a blank/newline)
.PP
\fIsanitize\fP converts any possible line originating from a DOS (CR/LF) or
MAC (CR) or unix (LF) and 0-terminate it.

.SH "EXAMPLES"
Some examples drafted, based on the new API presented.
.PP
An example reading in a full table into a two dimensional mdarray2, and
adding 1 to each element:
.nf

    table *t = table_open(filename, 0);
    mdarray2 d2 = table_md2rc(t,0,0,0,0);
    table_close(t);
    ncols = table_ncols(t);
    nrows = table_nrows(t);
    
    for (int i=0; i<nrows; i++) 
      for (int j=0; j<ncols; j++)
        d2[i][j] += 1.0;             // d2[row][col]

.fi
Here is an example of reading the table line by line, without any parsing,
but removing comment lines. This can be done in line by line streaming mode,
not allocating space for the whole table, for which \fBmode=1\fP is needed:
.nf

    table *t = table_open(filename, 1);
    int nrows = 0;
    string s;
    
    while ( (s=table_line(t)) ) {
	if (iscomment(s)) continue;
        nrows++	
        printf("%s\\n", s);  
    }
    
    table_close(t);
    dprintf(0,"Read %d lines\\n",nrows);
    
.fi
Dealing (and preserving) comments while reading in the whole table:
.nf

    table *t = table_open1(filename, 0, 0);
    int nrows = table_nrows(t);
    int ncols = table_ncols(t);  // this triggers a column counter
    
    string *sp = table_comments(t);   // not implemented yet
    while (*sp)
      printf("%s\n",*sp++);

    for (int j=0; j<nrows; j++)
	real *rp = table_rowrp(t, j);
	for (int i=0; i<ncols; i++)
	    printf("%g ",rp[i]);
	printf("\n");

    table_close(t);

.fi

.SH "XSV"
An interesting package to deal with tables is the XSV program

.nf
cat AAPL.csv | xsv table | head -2
cat AAPL.csv | xsv slice -i 1 | xsv table
cat AAPL.csv | xsv slice -i 1 | xsv flatten
cat AAPL.csv | xsv count

.fi

.SH "PERFORMANCE"
Anecdotally comparing the table I/O routines with python can be found
in $NEMO/scripts/csh/tabstat.py, which seems to indicate the C code
is about 4 times faster than numpy.

.SH "DIAGNOSTICS"
Low-level catastrophies (eg, bad filenames, parsing errors, wrong delimiters)
generate messages via \fIerror(3NEMO)\fP.

.SH "SEE ALSO"
mdarray(3NEMO), nemoinp(3NEMO), burststring(3NEMO), fits(5NEMO), table(5NEMO), ascii(7)
.PP
.nf

https://github.com/BurntSushi/xsv
https://heasarc.gsfc.nasa.gov/docs/software/fitsio/c/c_user/cfitsio.html
https://www.gnu.org/software/gnuastro/manual/html_node/Tables.html
XSV:  https://github.com/BurntSushi/xsv

.fi

.SH "FILES"
.nf
.ta +2.0i
src/kernel/tab  	table.c 
.fi

.SH "AUTHOR"
Peter Teuben

.SH "UPDATE HISTORY"
.nf
.ta +1.5i +5.5i
xx-sep-88	V1.0 written	PJT
6-aug-92	documented get_Xtable functions  	PJT
1-sep-95	added iscomment()	PJT
12-jul-03	fixed reading large table bufferrng	PJT
aug-2020	designing new table system	Sathvik/PJT
5-may-2022	finalizing implementation of table2	PJT/Parker/Yuzhu
31-dec-2022	add sanitize() to 0-terminate any style text	PJT
.fi


src/kernel/tab/table.c
/* some utility routines for table manipulations.
 *
 * New table V2 I/O routines: - test2 -
 *
 *    table_open
 *    table_line
 *    table_line1
 *    table_close
 *
 *
 * ==== some deprecatable functions ====
 *
 * get_line (instr, line)		get a line from a stream and return len
 *					(dangerous: max line len unpecified)
 *
 * parse (linenr, line,dat,ndat)	parse a line into data items
 *
 * strinsert (a, b, n)			insert string 'b' into 'a' for 'n' chars
 *
 * iscomment(line)			is this line a blank or comment line?
 * 
 *    1-jan-04      get_line::  changed EOF to return -1, and empty line to 0
 */
 
#include <stdinc.h>
#include <ctype.h>
#include <table.h>
#include <extstring.h>
#include <mdarray.h>

#if !defined(HUGE)
#define HUGE 1e20
#endif

#ifndef MAX_LINELEN
#define MAX_LINELEN  16384
#endif

bool ispipe(stream instr)
{
  off_t try = lseek(fileno(instr), 0, SEEK_CUR);
  if (try < 0) return TRUE;
  return FALSE;
}

/*
 * insert a string 'b' into 'a' replacing the first 'n' positions into 'a'
 *      (see als hsh.h in ..../hermes/lib  --  PJT)
 */

void strinsert(char *a, char *b, int n)
{
	int    i, idiff, alen, blen;

	alen = strlen(a);
	blen = strlen(b);
	idiff = blen - n;

	dprintf(8,"strinsert  in: (%s)-(%s)-%d\n", a, b, n);
	if (idiff < 0)		/* shift a to the left */
		for (i = n; i <= alen; i++)
			a[i + idiff] = a[i];
	else if (idiff > 0)	/* shift a to the right */
		for (i = alen; i >= n; i--)
			a[i + idiff] = a[i];

	for (i = 0; i < blen; i++)
		a[i] = b[i];
	dprintf(8,"strinsert out: (%s)-(%s)-%d\n", a, b, n);
}


int get_line(stream instr, char *line)
{
	int  c, i=0;
#if 1	
	static int _first = TRUE;

	if (_first) {
	  warning("old get_line() is still used - it is being deprecated");
	  _first = FALSE;
	}
#endif
	
	for(;;) {
		c=getc(instr);
		if (c==EOF)
			return -1;	/* error code: EOF */
		else if (c=='\n')
			break;
		line[i++]=c;
		if (i>MAX_LINELEN) {
			warning("get_line: max linelen (%d) exceeded; return 0",MAX_LINELEN);
			return 0;
		}
	}
	line[i]=0;
	return strlen(line);
}

#define PARESC  '%'

	/* linenr: current line number for %0 parameter */
	/* line:   ascii string */
	/* dat:    space where to store parsed items */
	/* ndat:   maximum that can be filled */

void parse(int linenr, char *line, double *dat, int ndat)
{
    int i, k, idat;
    char number[100];

    i=0;			/* position in line */
    while (line[i] != '\0') {
        if (line[i]==PARESC) {  /* referenced a column ? */
            k = i+1;            /* point to start of column index */
            while (isdigit(line[k]))
                k++;
            dprintf (2,"%s PARESC: i=%d k=%d\n",line,i,k);
            k -= i;                 /* number of digits + 1 of column index */
            idat = atoi(&line[i+1]);     /* index to column in dat[] */
            if (idat==0)
                sprintf (number,"(%d)",linenr);
            else if (idat <= ndat)
                sprintf (number,"(%16.10e)",dat[idat-1]); 
            else
                error("tabmath: referencing column %d\n",idat);
            strinsert(&line[i],number,k);
            i += strlen(number);
        } else
            i++;
    }
}


/*
 * iscomment: is a line a comment or blank line ?
 *		comment lines must start with '#' ';' '!' '/'
 *		or a NULL or newline '\n' (blank line)
 */

int iscomment(char *line)
{
    char *cp;

    /* first do a quick check on the first character */    
    if (*line=='#' || *line==';' || *line=='!' || *line=='/' || *line=='\n' || *line=='\0')
        return 1;
    /* then loop over the line until first non-whitespace is seen */
    for (cp=line; *cp; cp++)
        if (!isspace(*cp)) return 0;
    /* if none found, it bumped into the end of line and it's a comment */
    return 1;
}

/*
 * sanitize: convert to unix style 0 terminated lines
 *           if the line ending was DOS or MAC or UNIX
 *   MAC:   \r     (CR)
 *   UNIX:  \n     (LF)
 *   DOS:   \r\n   (CR-LF)
 */

void sanitize(char *line)
{
  int ret = strlen(line);
  //dprintf(1,"sanitize-1[%d]: '%s'\n", ret, line);
  if (line[ret-1] == '\r') {    //MAC
    line[ret-1] = '\0';
    ret--;
  }
  if (line[ret-1] == '\n') {    //UNIX
    line[ret-1] = '\0';
    ret--;
  }
  if (line[ret-1] == '\r') {    //DOS
    line[ret-1] = '\0';
    ret--;
  }
  //dprintf(1,"sanitize-2[%d]: '%s'\n", ret, line);
}

/*
 * parse_select, with dynamic reallocation
 *
 */

#if 0

void parse_select (select_str, select, nselect, Qsel, nQsel)
char *select_str;
int  **select, *nselect, nQsel;
bool **Qsel;
{
    error ("parse_select: Not implemented yet");
}

#endif


struct LineNode {        // simple linked list while reading lines
  char *val;
  struct LineNode* next;
};


table *table_open(stream instr, int mode)
{
  tableptr tptr = (tableptr) allocate(sizeof(table));
  dprintf(1,"table_open: mode=%d\n",mode);

  tptr->str     = instr;
  tptr->mode    = mode;
  tptr->lines   = NULL;
  tptr->nr      = 0;
  tptr->nc      = 0;
  tptr->linelen = 0;
  tptr->line    = NULL;
  dprintf(1,"table_open - got %d chars allocated at the start\n", tptr->linelen);

  if (mode <= 0) {   //  read table in memory, also separate header (comments) from body of table
    // note:    mode<0 treats all lines the same
    //          mode=0 should split comments out @todo
    dprintf(1,"linked list reading of table\n");

    // use a linked list to read all lines
    struct LineNode* first = (struct LineNode*) allocate(sizeof(struct LineNode));
    struct LineNode* curr = first;

    int nLines = 0;
    string line;

    while (1) {
      line = table_line(tptr);
      if (line == NULL)
	break;
      if (iscomment(line)) {
	// for now, skip them
	continue;
      }
      nLines++;
      curr->val = strdup(line);
      curr->next = (struct LineNode*) allocate(sizeof(struct LineNode));
      curr = curr->next;
      curr->next = NULL;
    }
    // allocate the array of nLines strings, point each line there
    tptr->nr = nLines;
    tptr->lines = (string *) allocate(nLines*sizeof(string));
    curr = first;
    for (int i=0; i<nLines; i++) {
      tptr->lines[i] = curr->val;
      curr = curr->next;
    }
    // @todo free up space of the linked list itself
  }
  
  // done!

  dprintf(1,"Read %d lines so far\n",tptr->nr);
  return tptr;
}

// seekable file mode=1 where you know number of lines
table *table_open1(stream instr, int mode, int nlines)
{
  tableptr tptr = (tableptr) allocate(sizeof(table));

  tptr->str   = instr;
  tptr->mode  = mode;
  tptr->lines = NULL;
  tptr->nr    = 0;
  tptr->nc    = 0;

  tptr->linelen = 0;
  tptr->line    = NULL;
  dprintf(0,"table_open - got %d chars allocated at the start\n", tptr->linelen);

  if (mode == 1)
    return tptr;

  // @todo what if nlines was not enough.
  // @todo skip comments here too?
  warning("table_open1: mode=%d nlines=%d",mode,nlines);
  tptr->nr = nlines;
  tptr->lines = (string *) allocate(nlines*sizeof(string));
  for (int i=0; i<nlines; i++)
    tptr->lines[i] = strdup(table_line(tptr));

  return tptr;
}

size_t table_nrows(tableptr tptr)
{
  return tptr->nr;
}


size_t table_ncols(tableptr tptr)
{
  // if tptr->nr > 0 and nc==0, force to read a line in mode=0 and set nc
  if (tptr->nr > 0 && tptr->nc == 0) {
    if (tptr->mode == 0) {
      string *sp = table_rowsp(tptr,0);
      dprintf(1,"table_ncols: processed first line to get nc -> %d\n",tptr->nc);
      tptr->nc = xstrlen(sp,sizeof(string)) - 1;
      // free
    } else {
      warning("mode=1 ... does not have ncols yet");
    }
  }
  return tptr->nc;
}


void table_reset(tableptr tptr)
{
  // reset the table   --    TBD
}


void table_close(tableptr tptr)
{
  // free that memory
  free(tptr->line);
  tptr->linelen = 0;
  // @todo - free more
}

/*
 *   table_line returns a NULL terminated line
 *   should be able to read unix (\n)   dos (\r\n) or mac (\r) files
 */
string table_line(tableptr tptr)
{
  ssize_t ret = getline(&(tptr->line), &(tptr->linelen), tptr->str);
  if (ret >= 0) {
    if (ret > 0) sanitize(tptr->line);
    return tptr->line;
  }
  // end of file
  return NULL;
}

table *table_cat(table* t1, table* t2, int mode)
{
  tableptr tptr = (tableptr) allocate(sizeof(table));
  
  tptr->lines = NULL;
  tptr->nr = 0;
  tptr->nc = 0;

#if 0
    tptr->linelen = 0;
    tptr->line = NULL;
#else
    tptr->linelen = 20;
    tptr->line = malloc(tptr->linelen);
#endif 
  dprintf(0,"table_cat - got %d chars allocated at the start\n", tptr->linelen);

  //concatenation(above and below)
  if(mode == 0) {
    //updating # of rows in the new table
    int totalLines = t1->nr + t2->nr;
    tptr->nr = totalLines;

    //updating # of columns to be the larger of the original two tables
    if(t1->nc > t2->nc){
      tptr->nc = t1->nc;
    } else {
      tptr->nc = t2->nc;
    }

    tptr->lines = (string*) allocate((tptr->nr) * sizeof(string));
    for(int i = 0; i < t1->nr; i++){
      tptr->lines[i] = strdup(table_line(t1));
    }
    for(int i = 0; i < t2->nr; i++){
      tptr->lines[i + t1->nr] = strdup(table_line(t2));
    }

  } else { //paste mode
    //assume number of rows are the same
    tptr->nr = t1->nr;
    //How do I update #of column in this case
    tptr->nc = t1->nc + t2->nc;
    tptr->lines = (string*) allocate((tptr->nr) * sizeof(string));

    for(int i = 0; i < t1->nr; i++){
      //@TODO: add white spaces between two lines
      tptr->lines[i] = strcat(strdup(table_line(t1)), strdup(table_line(t2)));
    }

  }

  return tptr;
} 


ssize_t table_line1(tableptr tptr, char **line, size_t *linelen, int newline)
{
  // in simple (streaming) mode, just get the next line
  if (tptr->mode == 1) {
    ssize_t len1 = getline(line, linelen, tptr->str);
    if (newline) return len1;
    if (*line[len1-1] == '\n') {
      line[len1-1] = '\0';
      len1--;
      return len1;
    }
  }

  error("line1: Unsupported table mode=%d", tptr->mode);
  return -1;
}


string table_row(tableptr tptr, int row)
{
  return tptr->lines[row];
}

// linked list placeholder
typedef struct lls {
  char *val;
  struct lls *next;
} lls;

// return list of zero terminated (extstring) pointers to the words in a row
string *table_rowsp(table *t, int row)
{
  char *line = strdup(t->lines[row]);
  int ntok = 0;
  char *token = strtok(line," ,\t");
  lls *first = (lls *) allocate(sizeof(lls));
  lls *curr = first;
  string *sp;
  int i;

  // find tokens
  while (token != NULL) {
    dprintf(2,"token: %d  %s\n",ntok, token);
    curr->val = token;
    curr->next = (lls *) allocate(sizeof(lls));
    curr = curr->next;
    curr->next = NULL;

    ++ntok;
    token = strtok(NULL, " ,\t");
  }

  // allocate the array to hold the tokens
  sp = (string *) allocate((ntok+1) * sizeof(string));
  curr = first;
  for (i=0; i<ntok; i++) {
    sp[i] = curr->val;
    curr = curr->next;
  }
  sp[ntok] = NULL;

  // consistency check on # columns
  if (t->nc == 0) {
    dprintf(1,"table_rowsp[%d]: setting ncols=%d\n",row,ntok);
    t->nc = ntok;
  } else if (ntok != t->nc)
    error("column number changed:  %d -> %d\n",t->nc, ntok);
  
  // free memory of linked list
  curr = first;
  while (curr->next) {
    dprintf(2,"free up LLS with %s\n",curr->val);
    first = curr->next;
    free(curr);
    curr = first;
  }
  // list of strings
  return sp;
}

/*
 *    a_i_j   row-major if a11 and a12 follow in memory    C,python:    data[0][0], data[0][1], ...
 *            col-major if a11 and a21 follow in memory    Fortran:     data(1,1), data(2,1)
 */

//
// nrow=0 and/or ncol=0 are allowed. It simply means all rows/cols are
// used in order
// Returns data[row][col]
//
mdarray2 table_md2rc(table *t, int nrow, int *rows, int ncol, int *cols)
{
  int nr = table_nrows(t);
  int nc = table_ncols(t);
  dprintf(1,"table_md2rc: table %d x %d \n",nr,nc);
  dprintf(1,"table_md2rc: data2 ncol=%d nrow=%d\n",ncol,nrow);
  mdarray2 a = allocate_mdarray2(nr,nc);    // a[nr][nc]
  int i,j;

  for (i=0; i<nr; i++) {
    string s = table_row(t,i);
    dprintf(1,"%d: %s\n",i,s);
    string *sp = table_rowsp(t,i);
    for (j=0; j<nc; j++) {
      a[i][j] = atof(sp[j]);
    }
  }

  return a;
}

// return a data[col][row] based table
// this is the more practical
// note column 0 has a special meaning, it's the row number
mdarray2 table_md2cr(table *t, int ncol, int *cols, int nrow, int *rows)
{
  int i,j,jidx;
  int nr = table_nrows(t);
  int nc = table_ncols(t);
  dprintf(1,"table_md2cr: table %d x %d \n",nr,nc);
  dprintf(1,"table_md2cr: data2 ncol=%d nrow=%d\n",ncol,nrow);
  if (ncol > 0) {
    for (j=0; j<ncol; j++) {
      jidx = cols[j];
      if (jidx < 0)  error("illegal column reference %d < 0", jidx);
      if (jidx > nc) error("illegal column reference %d > %d", jidx,nc);
    }
  }
  
  if (ncol>0) nc=ncol;
  if (nrow>0) nr=nrow;   // not supported yet  
  mdarray2 a = allocate_mdarray2(nc,nr);                  // a[nc][nr]

  for (i=0; i<nr; i++) {
    string s = table_row(t,i);
    dprintf(1,"%d: %s\n",i,s);
    string *sp = table_rowsp(t,i);
    for (j=0; j<nc; j++) {
      jidx = (ncol == 0 ?  j  :  cols[j]-1);
      if (jidx < 0)
	a[j][i] = i+1;
      else
	a[j][i] = atof(sp[jidx]);
    }
  }

  return a;
}


#ifdef TESTBED

#include <getparam.h>

string defv[] = {
    "file=???\n           Input or Output file",
    "write=f\n            Write mode? or else Write (w)",
    "test=0\n             testmode",
    "mode=0\n             buffering mode for table_open",
    "VERSION=2.0\n        29-apr-2022 PJT",
    NULL,
};

string usage = "testing tables";

void testmode1();
void testmode2();
void testmode3();
void testmode4();
void testmode5();

void nemo_main()
{
    int testmode = getiparam("test");
    int mode = getiparam("mode");
    tableptr tp1;
    stream instr, outstr;

    if (testmode == 1) return testmode1();
    if (testmode == 2) return testmode2();
    if (testmode == 3) return testmode3();
    if (testmode == 4) return testmode4();
    if (testmode == 5) return testmode5();

    warning("Falling through testmodes");

    if (getbparam("write")) {
      dprintf(0,"write mode\n");
      outstr = stropen(getparam("file"),"w");
      strclose(outstr);
      return;
    } 


    dprintf(0,"read mode\n");
    instr = stropen(getparam("file"),"r");
#if 0
    // original C
    size_t linelen = 0;                  // getline() is allowed to start from 0
    char *line = NULL;
    while (getline(&line, &linelen, instr) >= 0)
      printf("line[%ld] = %s\n", linelen, line);
    free(line);
#else
    // new style table2 : mode picks the reading by line, or using the
    //                    slightly slower linked-list reader
    char *cp;
    int nl = 0;
    tp1 = table_open(instr, mode);
    while ( (cp=table_line(tp1)) )
      nl++;
    table_close(tp1);
    dprintf(0,"Read %d extra in mode %s\n",nl,mode);
#endif

}



// linear list, but not supported in a pipe
void testmode1()
{
  string input = getparam("file");
  int nlines = nemo_file_lines(input,0);            // this will not work on pipes
  dprintf(0,"nemo_file_lines: %d\n",nlines);
  stream instr = stropen(input,"r");
  if (ispipe(instr))
    error("This mode is not supported with a pipe");
  rewind(instr);
    
  tableptr tp1 = table_open1(instr, 0, nlines);     // read the whole file in memory
  dprintf(0,"found nlines: %d\n",tp1->nr);
  printf("last  line: %s\n",table_row(tp1,tp1->nr - 1));
  printf("first line: %s\n",table_row(tp1,0));
}

// streaming
void testmode3()
{
  string input = getparam("file");
  string s;
  int nlines = nemo_file_lines(input,0);            // this will not work on pipes
  dprintf(0,"nemo_file_lines: %d\n",nlines);
  stream instr = stropen(input,"r");
  if (ispipe(instr))
    error("This mode is not supported with a pipe");
  rewind(instr);
    
  tableptr tp1 = table_open1(instr, 1, nlines);     // read the whole file in memory
  int nl=0;
  while ((s=table_line(tp1)))
    nl++;
    
  dprintf(0,"found nlines: %d\n",nl);
  // note cannot use table_row() since lines have not been saved

}

// linked list read, supports pipes
void testmode2()
{
  string input = getparam("file");
  stream instr = stropen(input,"r");
  tableptr tp1 = table_open(instr, 0);     // read the whole file in memory - linked list

  dprintf(0,"nlines: %d\n",tp1->nr);

  // assuming the linked list has been converted....
  printf("first line: %s\n",table_row(tp1,0));
  printf("last  line: %s\n",table_row(tp1,tp1->nr - 1));
	 
}

// linked list, streaming one line, and not keeping a buffer
void testmode4()
{
  string input = getparam("file");
  string s;
  stream instr = stropen(input,"r");
    
  tableptr tp1 = table_open(instr, 1);
  int nl=0;
  while ((s=table_line(tp1)))
    nl++;
    
  dprintf(0,"found nlines: %d\n",nl);
  // note cannot use table_row() since lines have not been saved
  dprintf(0,"has buffered nlines: %d\n",tp1->nr);
}

void testmode5()
{
  string input = getparam("file");
  stream instr = stropen(input,"r");
  tableptr tp1 = table_open(instr, 0);
  
#if 1
  int nl = table_nrows(tp1);
  int nc = table_ncols(tp1);
  dprintf(0,"testmode5: table %d x %d\n",nl,nc);

  for (int j=0; j<nl; j++) {
    string *sp = table_rowsp(tp1,j);
    int nsp = xstrlen(sp, sizeof(string))-1;
    if (nc == 0)
      nc = nsp;
    else if (nsp != nc)
      warning("Number of columns changed in row %d.... %d -> %d",j+1,nc,nsp);
    dprintf(1,"%d",j);
    for (int i=0; i<nsp; i++)
      dprintf(1," %s",sp[i]);
    dprintf(1,"\n");
    // free this memory is painful, so table should keep it.....
    char *cp = sp[0];
    free(cp);
  }
#endif
  
  mdarray2 a = table_md2rc(tp1,0,0,0,0);    //   a[row][col]
  
  dprintf(0,"after table_md2rc: table %d x %d\n",tp1->nr,tp1->nc);  
  for (int j=0; j<tp1->nr; j++) {
    dprintf(0,"%d",j);      
    for (int i=0; i<tp1->nc; i++)
      dprintf(0," %g",a[j][i]);
    dprintf(0,"\n");
  }

  mdarray2 b = table_md2cr(tp1,0,0,0,0);    //   a[col][row]
  
  dprintf(0,"after table_md2cr: table %d x %d\n",tp1->nr,tp1->nc);  
  for (int j=0; j<tp1->nr; j++) {
    dprintf(0,"%d",j);      
    for (int i=0; i<tp1->nc; i++)
      dprintf(0," %g",b[i][j]);
    dprintf(0,"\n");
  }
  
    
}

#endif
